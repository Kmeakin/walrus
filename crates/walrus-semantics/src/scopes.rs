use crate::{builtins::Builtin, diagnostic::Diagnostic, hir::*};
use arena::{Arena, ArenaMap, Idx};
use std::collections::HashMap;

pub fn scopes(module: &Module) -> Scopes {
    let mut scopes = Scopes::new();
    scopes.module_scope(&module.hir);
    scopes
}

pub type ScopeId = Idx<Scope>;
pub type Denotations = HashMap<Var, Denotation>;
type Vars = HashMap<Var, VarId>;
type LambdaDepth = u32;

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum Denotation {
    Local(VarId),
    Fn(FnDefId),
    Struct(StructDefId),
    Enum(EnumDefId),
    Builtin(Builtin),
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Scope {
    pub parent: Option<ScopeId>,
    pub denotations: Denotations,
    pub lambda_depth: LambdaDepth,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Scopes {
    scopes: Arena<Scope>,
    scope: ScopeId,
    scope_of_expr: ArenaMap<ExprId, ScopeId>,
    scope_of_type: ArenaMap<TypeId, ScopeId>,
    scope_of_pat: ArenaMap<PatId, ScopeId>,
    scope_of_var: ArenaMap<VarId, ScopeId>,
    pub diagnostics: Vec<Diagnostic>,
}

/// Public interface
impl Scopes {
    pub fn scope_chain(&self, scope: ScopeId) -> impl Iterator<Item = ScopeId> + '_ {
        std::iter::successors(Some(scope), move |&scope| self.scopes[scope].parent)
    }

    fn lookup_in_scope(&self, scope: ScopeId, var: &Var) -> Option<Denotation> {
        self.scope_chain(scope)
            .find_map(|scope| self.scopes[scope].denotations.get(var))
            .copied()
            .or_else(|| Builtin::lookup(var).map(Denotation::Builtin))
    }

    pub fn lookup_var(&self, id: VarId, var: &Var) -> Option<Denotation> {
        let scope = self.scope_of_var[id];
        self.lookup_in_scope(scope, var)
    }

    pub fn scope_of_var(&self, id: VarId) -> &Scope { &self.scopes[self.scope_of_var[id]] }
    pub fn scope_of_expr(&self, id: ExprId) -> &Scope { &self.scopes[self.scope_of_expr[id]] }
    pub fn scope_of_type(&self, id: TypeId) -> &Scope { &self.scopes[self.scope_of_type[id]] }
    pub fn scope_of_pat(&self, id: PatId) -> &Scope { &self.scopes[self.scope_of_pat[id]] }
}

/// Private helpers
impl Scopes {
    fn new() -> Self {
        let mut scopes = Arena::default();
        let scope = scopes.alloc(Scope::default());
        Self {
            scopes,
            scope,
            scope_of_expr: ArenaMap::default(),
            scope_of_type: ArenaMap::default(),
            scope_of_pat: ArenaMap::default(),
            scope_of_var: ArenaMap::default(),
            diagnostics: Vec::new(),
        }
    }

    fn set_scope_of_expr(&mut self, id: ExprId, scope: ScopeId) {
        self.scope_of_expr.insert(id, scope);
    }

    fn set_scope_of_type(&mut self, id: TypeId, scope: ScopeId) {
        self.scope_of_type.insert(id, scope);
    }

    fn set_scope_of_pat(&mut self, id: PatId, scope: ScopeId) {
        self.scope_of_pat.insert(id, scope);
    }

    fn set_scope_of_var(&mut self, id: VarId, scope: ScopeId) {
        self.scope_of_var.insert(id, scope);
    }

    fn child_scope(&mut self, scope: ScopeId, lambda_depth: LambdaDepth) -> ScopeId {
        self.scopes.alloc(Scope {
            parent: Some(scope),
            denotations: Denotations::new(),
            lambda_depth,
        })
    }

    fn in_child_scope(&mut self, mut f: impl FnMut(&mut Self)) {
        let old_scope = self.scope;
        self.scope = self.child_scope(old_scope, self.scopes[old_scope].lambda_depth);
        f(self);
        self.scope = old_scope;
    }

    fn in_child_lambda_scope(&mut self, mut f: impl FnMut(&mut Self)) {
        let old_scope = self.scope;
        self.scope = self.child_scope(old_scope, self.scopes[old_scope].lambda_depth + 1);
        f(self);
        self.scope = old_scope;
    }

    fn insert_denotation(
        &mut self,
        hir: &HirData,
        vars: &mut Vars,
        id: VarId,
        denotation: Denotation,
    ) {
        let var = hir[id].clone();
        if self.insert_var(hir, vars, id) {
            self.scopes[self.scope].denotations.insert(var, denotation);
        }
    }

    fn insert_var(&mut self, hir: &HirData, vars: &mut Vars, id: VarId) -> bool {
        self.set_scope_of_var(id, self.scope);
        let var = hir[id].clone();
        match vars.get(&var) {
            None => {
                vars.insert(var, id);
                true
            }
            Some(first) => {
                self.diagnostics.push(Diagnostic::DuplicateVar {
                    first: *first,
                    second: id,
                });
                false
            }
        }
    }
}

/// The actual implementation
impl Scopes {
    fn module_scope(&mut self, hir: &HirData) {
        let mut toplevel_defs = Vars::new();

        for decl in hir.decls() {
            match decl {
                Decl::Struct(id) => self.struct_def_scope(hir, &mut toplevel_defs, id),
                Decl::Enum(id) => self.enum_def_scope(hir, &mut toplevel_defs, id),
                Decl::Fn(id) => self.fn_def_scope(hir, &mut toplevel_defs, id),
            }
        }
    }

    fn struct_def_scope(&mut self, hir: &HirData, toplevel_defs: &mut Vars, id: StructDefId) {
        let struct_def = &hir[id];
        self.insert_denotation(hir, toplevel_defs, struct_def.name, Denotation::Struct(id));

        let mut fields = Vars::new();
        for field in &struct_def.fields {
            self.insert_var(hir, &mut fields, field.name);
            self.type_scope(hir, field.ty);
        }
    }

    fn enum_def_scope(&mut self, hir: &HirData, toplevel_defs: &mut Vars, id: EnumDefId) {
        let enum_def = &hir[id];
        self.insert_denotation(hir, toplevel_defs, enum_def.name, Denotation::Enum(id));

        let mut variants = Vars::new();
        for variant in &enum_def.variants {
            self.insert_var(hir, &mut variants, variant.name);
            let mut fields = Vars::new();
            for field in &variant.fields {
                self.insert_var(hir, &mut fields, field.name);
                self.type_scope(hir, field.ty);
            }
        }
    }

    fn fn_def_scope(&mut self, hir: &HirData, toplevel_defs: &mut Vars, id: FnDefId) {
        let fn_def = &hir[id];
        self.insert_denotation(hir, toplevel_defs, fn_def.name, Denotation::Fn(id));

        self.in_child_scope(|this| {
            let mut params = Vars::new();
            for param in &fn_def.params {
                this.param_scope(hir, &mut params, param)
            }
            if let Some(ty) = fn_def.ret_type {
                this.type_scope(hir, ty)
            }
            this.expr_scope(hir, fn_def.expr)
        })
    }

    fn param_scope(&mut self, hir: &HirData, vars: &mut Vars, param: &Param) {
        let Param { pat, ty } = param;
        self.pat_scope(hir, vars, *pat);
        if let Some(ty) = ty {
            self.type_scope(hir, *ty)
        }
    }

    fn pat_scope(&mut self, hir: &HirData, vars: &mut Vars, id: PatId) {
        self.set_scope_of_pat(id, self.scope);
        let pat = &hir[id];
        match pat {
            Pat::Var { var, .. } => {
                self.insert_denotation(hir, vars, *var, Denotation::Local(*var))
            }
            Pat::Struct { name, fields, .. } | Pat::Enum { name, fields, .. } => {
                self.set_scope_of_var(*name, self.scope);
                let mut seen_fields = Vars::new();
                for field in fields {
                    self.set_scope_of_var(field.name, self.scope);
                    self.insert_var(hir, &mut seen_fields, field.name);

                    match field {
                        FieldPat { pat: Some(pat), .. } => self.pat_scope(hir, vars, *pat),
                        FieldPat { name, pat: None } => {
                            self.insert_denotation(hir, vars, *name, Denotation::Local(*name))
                        }
                    }
                }
            }
            pat => pat.walk_child_pats(|id| self.pat_scope(hir, vars, id)),
        }
    }

    fn type_scope(&mut self, hir: &HirData, id: TypeId) {
        self.set_scope_of_type(id, self.scope);
        let ty = &hir[id];
        match ty {
            Type::Var(var) => self.set_scope_of_var(*var, self.scope),
            ty => ty.walk_child_types(|id| self.type_scope(hir, id)),
        }
    }

    fn expr_scope(&mut self, hir: &HirData, id: ExprId) {
        self.set_scope_of_expr(id, self.scope);
        let expr = &hir[id];
        match expr {
            Expr::Block { stmts, expr } => {
                self.in_child_scope(|this| this.block_scope(hir, stmts, *expr))
            }
            Expr::Lambda { params, expr } => self.in_child_lambda_scope(|this| {
                let mut vars = Vars::new();
                for param in params {
                    this.param_scope(hir, &mut vars, param)
                }
                this.expr_scope(hir, *expr)
            }),
            Expr::Match { test, cases } => {
                self.expr_scope(hir, *test);
                for case in cases {
                    self.in_child_scope(|this| {
                        let mut vars = Vars::new();
                        this.pat_scope(hir, &mut vars, case.pat);
                        this.expr_scope(hir, case.expr)
                    })
                }
            }
            Expr::Var(var) => self.set_scope_of_var(*var, self.scope),
            Expr::Struct { name, fields } | Expr::Enum { name, fields, .. } => {
                self.set_scope_of_var(*name, self.scope);
                let mut seen_fields = Vars::new();
                for field in fields {
                    self.insert_var(hir, &mut seen_fields, field.name);
                    self.set_scope_of_var(field.name, self.scope);
                    self.expr_scope(hir, field.val)
                }
            }
            expr => expr.walk_child_exprs(|id| self.expr_scope(hir, id)),
        }
    }

    fn block_scope(&mut self, hir: &HirData, stmts: &[Stmt], last: Option<ExprId>) {
        match stmts {
            [Stmt::Expr(expr), rest @ ..] => {
                self.expr_scope(hir, *expr);
                self.block_scope(hir, rest, last);
            }
            [Stmt::Let { pat, ty, expr }, rest @ ..] => self.in_child_scope(|this| {
                this.pat_scope(hir, &mut Vars::new(), *pat);
                if let Some(ty) = ty {
                    this.type_scope(hir, *ty)
                }
                this.expr_scope(hir, *expr);

                this.block_scope(hir, rest, last)
            }),
            [] => {
                if let Some(expr) = last {
                    self.expr_scope(hir, expr)
                }
            }
        }
    }
}
