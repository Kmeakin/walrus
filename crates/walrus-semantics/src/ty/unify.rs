use super::*;
use ena::unify::{InPlaceUnificationTable, NoError, UnifyKey, UnifyValue};

#[derive(Clone, Debug, Default)]
pub struct InferenceTable {
    pub(super) var_unification_table: InPlaceUnificationTable<TypeVarId>,
}

/// The ID of a type variable.
#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub struct TypeVarId(pub(super) u32);

impl UnifyKey for TypeVarId {
    type Value = TypeVarValue;
    fn index(&self) -> u32 { self.0 }
    fn from_index(i: u32) -> Self { Self(i) }
    fn tag() -> &'static str { "TypeVarId" }
}

/// The value of a type variable: either we already know the type, or we don't
/// know it yet.
#[derive(Clone, PartialEq, Eq, Debug)]
pub enum TypeVarValue {
    Known(Type),
    Unknown,
}

impl TypeVarValue {
    const fn as_known(&self) -> Option<&Type> {
        match self {
            Self::Known(ty) => Some(ty),
            Self::Unknown => None,
        }
    }
}

impl UnifyValue for TypeVarValue {
    type Error = NoError;

    fn unify_values(value1: &Self, value2: &Self) -> Result<Self, NoError> {
        match (value1, value2) {
            // We should never equate two type variables, both of which have
            // known types. Instead, we recursively equate those types.
            (Self::Known(t1), Self::Known(t2)) => panic!(
                "equating two type variables, both of which have known types: {:?} and {:?}",
                t1, t2
            ),

            // If one side is known, prefer that one.
            (Self::Known(..), Self::Unknown) => Ok(value1.clone()),
            (Self::Unknown, Self::Known(..)) => Ok(value2.clone()),

            (Self::Unknown, Self::Unknown) => Ok(Self::Unknown),
        }
    }
}

impl InferenceTable {
    pub fn new_type_var(&mut self) -> TypeVarId {
        self.var_unification_table.new_key(TypeVarValue::Unknown)
    }

    /// Propagates the type completely; type variables without known type are
    /// replaced by `Type::Unknown`.
    pub(crate) fn propagate_type_completely(&mut self, ty: &Type) -> Type {
        self.propagate_type_completely_inner(&mut Vec::new(), ty)
    }

    pub(crate) fn propagate_type_completely_inner(
        &mut self,
        tv_stack: &mut Vec<TypeVarId>,
        ty: &Type,
    ) -> Type {
        ty.clone().fold(&mut |ty| match ty {
            Type::Infer(tv) => {
                let inner = tv.to_inner();
                if tv_stack.contains(&inner) {
                    return tv.fallback_value();
                }
                self.var_unification_table
                    .inlined_probe_value(inner)
                    .as_known()
                    .map_or_else(
                        || tv.fallback_value(),
                        |known_ty| {
                            // known_ty may contain other variables that are known by now
                            tv_stack.push(inner);
                            let result = self.propagate_type_completely_inner(tv_stack, known_ty);
                            tv_stack.pop();
                            result
                        },
                    )
            }
            _ => ty,
        })
    }

    /// Propagates the type as far as currently possible, replacing type
    /// variables by their known types. All types returned by the `infer_*`
    /// functions should be propagated as far as possible, i.e. contain no
    /// type variables with known type.
    pub(crate) fn propagate_type_as_far_as_possible(&mut self, ty: &Type) -> Type {
        self.propagate_type_as_far_as_possible_inner(&mut Vec::new(), ty)
    }

    pub(crate) fn propagate_type_as_far_as_possible_inner(
        &mut self,
        tv_stack: &mut Vec<TypeVarId>,
        ty: &Type,
    ) -> Type {
        ty.clone().fold(&mut |ty| match ty {
            Type::Infer(tv) => {
                let inner = tv.to_inner();
                if tv_stack.contains(&inner) {
                    return tv.fallback_value();
                }
                self.var_unification_table
                    .inlined_probe_value(inner)
                    .as_known()
                    .map_or(ty, |known_ty| {
                        tv_stack.push(inner);
                        let result =
                            self.propagate_type_as_far_as_possible_inner(tv_stack, known_ty);
                        tv_stack.pop();
                        result
                    })
            }
            _ => ty,
        })
    }

    fn propagate_type_shallow(&mut self, ty: &Type) -> Type {
        match ty {
            Type::Infer(tv) => {
                let inner = tv.to_inner();
                let value = self.var_unification_table.inlined_probe_value(inner);
                match value.as_known() {
                    Some(known_ty) => known_ty.clone(),
                    None => ty.clone(),
                }
            }
            _ => ty.clone(),
        }
    }

    pub fn unify(&mut self, t1: &Type, t2: &Type) -> bool {
        if t1 == t2 {
            return true;
        }

        let t1 = self.propagate_type_shallow(&t1.clone());
        let t2 = self.propagate_type_shallow(&t2.clone());
        match (t1, t2) {
            (
                Type::App {
                    ctor: ctor1,
                    params: params1,
                },
                Type::App {
                    ctor: ctor2,
                    params: params2,
                },
            ) if ctor1 == ctor2 => {
                params1.len() == params2.len()
                    && (params1.iter())
                        .zip(params2.iter())
                        .all(|(t1, t2)| self.unify(t1, t2))
            }
            (Type::Unknown, _) | (_, Type::Unknown) => true,
            (Type::Infer(InferType::Var(tv1)), Type::Infer(InferType::Var(tv2))) => {
                self.var_unification_table.union(tv1, tv2);
                true
            }

            (Type::Infer(InferType::Var(tv)), other) | (other, Type::Infer(InferType::Var(tv))) => {
                self.var_unification_table
                    .union_value(tv, TypeVarValue::Known(other));
                true
            }

            // below Type::Infer so that
            // unify(TypeVar, Never) => Never, instead of
            // unify(TypeVar, Never) => TypeVar
            (ty, _) | (_, ty) if ty == Type::NEVER => true,

            _ => false,
        }
    }
}
