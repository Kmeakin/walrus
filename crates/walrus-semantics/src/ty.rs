use crate::scopes::BuiltinType;

use self::unify::TypeVarId;

mod infer;
mod unify;

pub use self::infer::infer;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type {
    Unknown,
    App(TypeApp),
    Infer(InferType),
}

impl Type {
    pub const UNIT: Self = Self::new0(TypeCtor::Tuple);
    pub const BOOL: Self = Self::new0(TypeCtor::Bool);
    pub const INT: Self = Self::new0(TypeCtor::Int);
    pub const FLOAT: Self = Self::new0(TypeCtor::Float);
    pub const CHAR: Self = Self::new0(TypeCtor::Char);
    pub const NEVER: Self = Self::new0(TypeCtor::Never);

    pub const fn new0(ctor: TypeCtor) -> Self { Self::App(TypeApp::new0(ctor)) }
    pub fn function(params: Vec<Type>, ret: Type) -> Self { Self::App(TypeApp::func(params, ret)) }
    pub fn tuple(tys: Vec<Type>) -> Self { Self::App(TypeApp::tuple(tys)) }
    pub fn as_tuple(&self) -> Option<&[Type]> {
        match self {
            Type::App(TypeApp {
                ctor: TypeCtor::Tuple,
                params,
            }) => Some(params),
            _ => None,
        }
    }
    pub fn as_fn(&self) -> Option<(&[Type], &Type)> {
        match self {
            Type::App(TypeApp {
                ctor: TypeCtor::Fn,
                params,
            }) => Some(
                params
                    .split_last()
                    .map(|(ret, params)| (params, ret))
                    .unwrap(),
            ),
            _ => None,
        }
    }

    pub fn is_num(&self) -> bool {
        match self {
            _ => [Self::INT, Self::FLOAT].contains(self),
        }
    }

    pub fn is_eq(&self) -> bool {
        match self {
            Type::App(TypeApp {
                ctor: TypeCtor::Tuple,
                params,
            }) => params.iter().all(Self::is_num),
            _ => [Self::BOOL, Self::INT, Self::FLOAT, Self::CHAR].contains(self),
        }
    }

    pub fn is_cmp(&self) -> bool {
        match self {
            Type::App(TypeApp {
                ctor: TypeCtor::Tuple,
                params,
            }) => params.iter().all(Self::is_num),
            _ => [Self::INT, Self::FLOAT].contains(self),
        }
    }

    fn walk_mut(&mut self, f: &mut impl FnMut(&mut Self)) {
        f(self);
        match self {
            Self::App(ty) => {
                for t in &mut ty.params {
                    t.walk_mut(f);
                }
            }
            Self::Infer(_) | Self::Unknown => {}
        }
    }

    fn fold(mut self, f: &mut impl FnMut(Self) -> Self) -> Self {
        self.walk_mut(&mut |ty_mut| {
            let ty = std::mem::replace(ty_mut, Self::Unknown);
            *ty_mut = f(ty);
        });
        self
    }
}

impl From<BuiltinType> for Type {
    fn from(ty: BuiltinType) -> Self {
        match ty {
            BuiltinType::Bool => Self::BOOL,
            BuiltinType::Int => Self::INT,
            BuiltinType::Float => Self::FLOAT,
            BuiltinType::Char => Self::CHAR,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TypeApp {
    pub ctor: TypeCtor,
    pub params: Vec<Type>,
}

impl TypeApp {
    pub const fn new0(ctor: TypeCtor) -> Self {
        Self {
            ctor,
            params: vec![],
        }
    }
    pub fn func(params: Vec<Type>, ret: Type) -> Self {
        Self {
            ctor: TypeCtor::Fn,
            params: params.into_iter().chain(Some(ret)).collect(),
        }
    }
    pub fn tuple(tys: Vec<Type>) -> Self {
        Self {
            ctor: TypeCtor::Tuple,
            params: tys,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TypeCtor {
    Bool,
    Int,
    Float,
    Char,
    Never,
    Tuple,
    Fn,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum InferType {
    Var(TypeVarId),
    /* TODO
    Int,
    Float,
    */
}

impl InferType {
    const fn fallback_value(self) -> Type {
        match self {
            Self::Var(_) => Type::Unknown,
        }
    }

    const fn to_inner(self) -> TypeVarId {
        match self {
            Self::Var(ty) => ty,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::*;

    macro_rules! test_infer {
        ($name:ident, $src:expr, $expected:expr) => {
            #[test]
            fn $name() { test_infer($src, $expected) }
        };
    }

    fn test_infer(src: &str, expected: Type) {
        let syntax = walrus_parser::parse(src);
        let hir = crate::hir::lower(&syntax);
        let scopes = crate::scopes::scopes(&hir);
        let types = infer(hir.clone(), scopes);

        let first_fn = &hir.data.fn_defs.iter().next().unwrap();
        let ret_type = &types.type_of_fn[first_fn.0].ret;

        assert_eq!(ret_type, &expected);

        let mut settings = insta::Settings::new();
        settings.set_snapshot_path("../snapshots");
        settings.set_prepend_module_to_snapshot(false);
        settings.bind(|| assert_debug_snapshot!(types));
    }

    test_infer!(empty_fn, r#"fn f() {}"#, Type::UNIT);
    test_infer!(infer_ret_type, r#"fn f() -> _ {}"#, Type::UNIT);
    test_infer!(bool_lit, r#"fn f() -> _ {true}"#, Type::BOOL);
    test_infer!(int_lit, r#"fn f() -> _ {1}"#, Type::INT);
    test_infer!(float_lit, r#"fn f() -> _ {1.0}"#, Type::FLOAT);
    test_infer!(char_lit, r#"fn f() -> _ {'a'}"#, Type::CHAR);

    test_infer!(unit_stmt, r#"fn f() -> _ {1;}"#, Type::UNIT);
    test_infer!(let_var, r#"fn f() -> _ {let x = 5; x}"#, Type::INT);
    test_infer!(
        fn_var,
        r#"
fn f() -> _ {g}
fn g() -> _ {1}
"#,
        Type::function(vec![], Type::INT)
    );

    test_infer!(
        if_then_else,
        r#"fn f() -> _ {if true {1} else {0}}"#,
        Type::INT
    );

    test_infer!(
        tuple,
        r#"fn f() -> _ {(1,false)}"#,
        Type::tuple(vec![Type::INT, Type::BOOL])
    );

    test_infer!(
        lambda,
        r#"fn f() -> _ { (x: Int) => false }"#,
        Type::function(vec![Type::INT], Type::BOOL)
    );

    test_infer!(
        lambda2,
        r#"fn f() -> _ { (x) => x + 1 }"#,
        Type::function(vec![Type::INT], Type::INT)
    );

    test_infer!(
        lambda_call,
        r#"
fn f() -> _ {
    let id = (x) => x;
    id(1)
}"#,
        Type::INT
    );

    test_infer!(unary_sub, r#"fn f() -> _ {-0}"#, Type::INT);
    test_infer!(unary_add, r#"fn f() -> _ {+0}"#, Type::INT);

    test_infer!(cmp, r#"fn f() -> _ {0 == 1}"#, Type::BOOL);
    test_infer!(loop_never, r#"fn f() -> _ { loop {} }"#, Type::NEVER);
    test_infer!(loop_unit, r#"fn f() -> _ { loop { break } }"#, Type::UNIT);
    test_infer!(loop_int, r#"fn f() -> _ { loop { break 1 } }"#, Type::INT);

    test_infer!(return_unit, r#"fn f() -> _ { return }"#, Type::UNIT);
    test_infer!(return_int, r#"fn f() -> _ { return 1 }"#, Type::INT);

    test_infer!(
        lambda_return_unit,
        r#"fn f() -> _ { () => return }"#,
        Type::UNIT
    );
    test_infer!(
        lambda_return_int,
        r#"fn f() -> _ { () => return 1 }"#,
        Type::INT
    );

    test_infer!(
        fn_params,
        r#"
fn f() -> _ { g }
fn g(_: Int) -> _ {}
"#,
        Type::function(vec![Type::INT], Type::UNIT)
    );

    test_infer!(
        factorial,
        r#"
fn f() -> _ { factorial }
fn factorial(x: _) -> _ {
    if x == 0 { 1 }
    else      { x * factorial(x-1) }
}
"#,
        Type::function(vec![Type::INT], Type::INT)
    );

    test_infer!(
        add,
        r#"
fn f() -> _ { add }
fn add(x: _, y: _) -> _ { x + y + 1 }
"#,
        Type::function(vec![Type::INT, Type::INT], Type::INT)
    );

    test_infer!(
        parity,
        r#"
fn f() -> _ { (is_odd, is_even) }
fn is_odd(x: _)  -> _ { if x == 0 { false } else { is_even(x - 1) } }
fn is_even(x: _) -> _ { if x == 0 { true  } else { is_odd(x - 1)  } }
"#,
        Type::tuple(vec![
            Type::function(vec![Type::INT], Type::BOOL),
            Type::function(vec![Type::INT], Type::BOOL),
        ])
    );
}
