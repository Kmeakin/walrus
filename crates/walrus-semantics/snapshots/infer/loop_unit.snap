---
source: crates/walrus-semantics/src/ty.rs
expression: types

---
InferenceResult {
    type_of_expr: {
        0: Primitive(
            Never,
        ),
        1: Primitive(
            Never,
        ),
        2: Tuple(
            [],
        ),
        3: Tuple(
            [],
        ),
    },
    type_of_type: {
        0: Tuple(
            [],
        ),
    },
    type_of_pat: {},
    type_of_fn: {
        0: FnType {
            params: [],
            ret: Tuple(
                [],
            ),
        },
    },
    diagnostics: [],
}
